sources/pam.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * Odyssey.
+ *
+ * Scalable PostgreSQL connection pooler.
+*/
+
+#include <security/pam_appl.h>
+#include <kiwi.h>
+
+struct auth_request {
+    char *password;
+};
+
+
+static int
+pam_conversation(int msgc,
+		const struct pam_message **msgv,
+	       	struct pam_response **rspv,
+		void *authdata) {
+	struct auth_request *request = (struct auth_request *) authdata;
+	int counter, retcode;
+	
+	if (msgc < 1 || msgv == NULL || request == NULL) {
+		return PAM_CONV_ERR;
+	}
+	*rspv = malloc(msgc * sizeof(struct pam_response));
+	if (*rspv == NULL) {
+		return PAM_CONV_ERR;
+	}
+	
+	memset(*rspv, 0, msgc * sizeof(struct pam_response));
+	retcode = PAM_SUCCESS;
+	
+	for (counter = 0; retcode == PAM_SUCCESS && counter < msgc; counter++) {
+		if (msgv[counter]->msg_style == PAM_PROMPT_ECHO_OFF) {
+			(*rspv)[counter].resp = strdup(request->password);
+			if ((*rspv)[counter].resp == NULL) {
+				retcode = PAM_CONV_ERR;
+			}
+		}
+	}
+	
+	if (retcode != PAM_SUCCESS) {
+		for (counter = 0; counter < msgc; counter++)
+			free((*rspv)[counter].resp);
+		free(*rspv);
+	}
+	return retcode;
+}
+
+static int
```

Comment:
do we need this function to return PAM statuses, maybe just common `-1, 0` would be more convenient?

Reply:
sure, fixed